### ISSUE TITLE
Go to references

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/vscode-terraform#651]
## Problem Statement

[Terraform allows referencing](https://www.terraform.io/docs/language/expressions/references.html) other pieces of config, or the data the config is referring to.

For example a variable can be referenced via `var.name`:

```hcl
variable "test" {
  default = "blah"
}

resource "aws_alb" "name" {
  count = var.test
}
```

Users would benefit from:
 - knowing how many references there are throughout the module to a given block (variable, local, resource, data source etc.)
 - being able to navigate to these references easily, e.g. from `variable "test"` block to any `var.test` occurrences.

### Code Lens (annotation w/ number of references)

This should be achievable via Code Lens which are part of the LSP:
https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_codeLens

This allows us to associate an "annotation" with a range of code and a command to be executed when the user clicks on the "annotation". In this case the code range may include the block header (block type + any labels).

### Go To References

This is also part of LSP:
https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_references

## User Impact

Referencing is an essential part of Terraform language that many users use, and so most users would likely benefit from this feature.

This feature was also part of the VS Code extension v1 version.

Users reported this via GitHub:
 - https://github.com/hashicorp/terraform-ls/issues/254

### Code Lens

User will see `N references` annotation above any block that is referenced at least once anywhere else in the module.

For more than 1 reference, user will be able to pick which reference to navigate to.

![2021-05-26 09 34 08](https://user-images.githubusercontent.com/287584/119636217-5206e280-be0c-11eb-9ee7-038e66aa55a7.gif)

### Go To References

User will be able to right-click to a block header and pick "Go to references", which will bring the IDE-native dialog for references.

![2021-05-26 12 15 32](https://user-images.githubusercontent.com/287584/119650830-2d1a6b80-be1c-11eb-9751-833419bb8db2.gif)

## Proposal

 - Introduce `CollectReferenceOrigins(filename string) (ReferenceOrigins, error)` to `hcl-lang` decoder
 - Persist reference origins in LS module state
 - Refactor `lang.Reference` to `lang.ReferenceTarget`
 - Introduce `ReferenceTargetAtPos(filename string, pos hcl.Pos) lang.ReferenceTarget`
 - Introduce `ReferenceOrigins.MatchingAddress(addr lang.Address) lang.ReferenceOrigins`
 - implement `textDocument/references` which will call `ReferenceTargetAtPos` and use the obtained `ReferenceTarget.Address` to filter out persisted reference origins in module state via `ReferenceOrigins.MatchingAddress`

```go
type ReferenceOrigin struct {
  Address lang.Address
  TraversalRange hcl.Range
}
```

## Blocking Questions

 - How do we implement the "quick-pick" dialog for the code lens, so that most of the logic can stay on the language server?

**Comment by radeksimko** [comment]
The "Go To References" part (where you can bring up a list of references via context menu) was implemented in language server 0.19.0 (released today) as part of https://github.com/hashicorp/terraform-ls/pull/572 and https://github.com/hashicorp/terraform-ls/pull/580

Code lens is currently being worked on.

**Comment by radeksimko** [comment]
This was implemented as part of https://github.com/hashicorp/vscode-terraform/pull/686 which was shipped in extension version `2.14.0` and https://github.com/hashicorp/terraform-ls/pull/584 which was shipped in language server version `0.20.0` (yesterday's release).

There are further planned improvements in this area - feel free to subscribe to any of these issues

 - https://github.com/hashicorp/vscode-terraform/issues/706
 - https://github.com/hashicorp/vscode-terraform/issues/670
 - https://github.com/hashicorp/terraform-ls/issues/558 / https://github.com/hashicorp/vscode-terraform/issues/669
 - https://github.com/hashicorp/terraform-ls/issues/581
 - https://github.com/hashicorp/terraform-ls/issues/583

or open a new one if the problem/feature isn't described in any of them.
